src/Components/Components/src/NavigationManager.cs

```diff
@@ -398,17 +392,30 @@ protected async ValueTask<bool> NotifyLocationChangingAsync(string uri, string?
         }
     }
 
+    private async ValueTask InvokeLocationChangingHandlerAsync(Func<LocationChangingContext, ValueTask> handler, LocationChangingContext context)
+    {
+        try
+        {
+            await handler(context);
+        }
+        catch (OperationCanceledException)
+        {
+            // Ignore exceptions caused by cancellations.
+        }
+        catch (Exception ex)
+        {
+            HandleLocationChangingHandlerException(ex, context);
+        }
+    }
+
     /// <summary>
-    /// Invokes the provided <paramref name="handler"/>, passing it the given <paramref name="context"/>.
-    /// This method can be overridden to analyze the state of the handler task even after
-    /// <see cref="NotifyLocationChangingAsync(string, string?, bool)"/> completes. For example, this can be useful for
-    /// processing exceptions thrown from handlers that continue running after the navigation ends.
+    /// Handles exceptions thrown in location changing handlers.
     /// </summary>
-    /// <param name="handler">The handler to invoke.</param>
-    /// <param name="context">The context to pass to the handler.</param>
-    /// <returns></returns>
-    protected virtual ValueTask InvokeLocationChangingHandlerAsync(Func<LocationChangingContext, ValueTask> handler, LocationChangingContext context)
-        => handler(context);
+    /// <param name="ex">The exception to handle.</param>
+    /// <param name="context">The context passed to the handler.</param>
+    protected virtual void HandleLocationChangingHandlerException(Exception ex, LocationChangingContext context)
+    {
+    }
```

Comment:
Would it work to have this throw by default? For example, `throw new InvalidOperation("NavigationManager subclasses must override HandleLocationChangingHandlerException in order to support navigation cancellation.")`

The intention is that subclasses are meant to override this if they will ever get into a situation where it might be called, isn't it? We don't want people to forget to do that and then just be ignoring exceptions without knowing about it.

I don't know whether throwing from here will result in the exception showing up, or whether it just gets swallowed above!

Reply:
Sure, this seems good. It's just not guaranteed that `NavigationManager` subclass implementors would see this exception, because in order to see it they would need to:
1. Wrap the call to `NotifyLocationChangingAsync()` in a try/catch and do something useful to surface the exception.
2. Test the case where exceptions are thrown from location changing handlers.

Furthermore, exceptions thrown from `HandleLocationChangingHandlerException()` won't get caught for handlers that throw after `NotifyLocationChangingAsync()` returns.

That said, this does still seem like an improvement over swallowing exceptions silently by default.